File: Deep/Deep/Features/DeepSession/Models/BreathEngine.swift
import Foundation
import Observation

/// Drives one guided session: which breath phase we're in, which round, and
/// whether the practice is paused or complete. Pure clockwork — no UI, no
/// audio — so the screen stays a projection of this state and previews can
/// pin any moment of the session.
@Observable @MainActor
final class BreathEngine {
  enum Phase {
    case inhale
    case exhale
    case finished
  }

  let session: DeepSession
  private(set) var phase: Phase = .inhale
  /// 1-based current round, capped at `session.cycles`.
  private(set) var cycle = 1
  private(set) var isPaused = false

  private var advanceTask: Task<Void, Never>?
  private var phaseEnds: Date?
  /// Seconds left in the current phase, captured when pausing.
  private var pausedRemaining: TimeInterval?

  init(session: DeepSession) {
    self.session = session
  }

  /// Starts the clock. Safe to call once from `onAppear`; later calls no-op.
  func begin() {
    guard advanceTask == nil, phase != .finished, !isPaused else { return }
    scheduleAdvance(after: duration(of: phase))
  }

  func togglePaused() {
    guard phase != .finished else { return }
    isPaused ? resume() : pause()
  }

  /// Stops the clock without changing visible state; call on dismiss.
  func cancel() {
    advanceTask?.cancel()
    advanceTask = nil
  }

  /// Seconds left in the current phase — live while the clock runs, frozen
  /// while paused. `nil` before `begin()` and once finished.
  var remainingInPhase: TimeInterval? {
    pausedRemaining ?? phaseEnds.map { max(0, $0.timeIntervalSinceNow) }
  }

  /// The full length of the phase we're currently in.
  var currentPhaseDuration: TimeInterval { duration(of: phase) }

  /// Seconds of practice still ahead — what remains of this phase, the rest of
  /// this round, and every round after it. Read live, like `remainingInPhase`,
  /// so a caller re-reads it whenever something else moves them.
  var remainingInSession: TimeInterval {
    guard phase != .finished else { return 0 }
    let inThisPhase = remainingInPhase ?? duration(of: phase)
    let restOfRound = phase == .inhale ? session.exhale : 0
    let roundsAhead = TimeInterval(max(0, session.cycles - cycle))
    return inThisPhase + restOfRound + roundsAhead * session.cycleDuration
  }

  /// Internal (not just `togglePaused`'s half) so the coordinator can settle
  /// the practice when the app leaves the foreground.
  func pause() {
    guard phase != .finished, !isPaused else { return }
    isPaused = true
    pausedRemaining = phaseEnds.map { max(0, $0.timeIntervalSinceNow) }
    cancel()
  }

  private func resume() {
    isPaused = false
    scheduleAdvance(after: pausedRemaining ?? duration(of: phase))
    pausedRemaining = nil
  }

  private func scheduleAdvance(after seconds: TimeInterval) {
    phaseEnds = Date(timeIntervalSinceNow: seconds)
    advanceTask = Task { [weak self] in
      try? await Task.sleep(for: .seconds(seconds))
      guard !Task.isCancelled else { return }
      self?.advance()
    }
  }

  private func advance() {
    switch phase {
    case .inhale:
      phase = .exhale
      scheduleAdvance(after: session.exhale)
    case .exhale:
      if cycle >= session.cycles {
        phase = .finished
        advanceTask = nil
      } else {
        cycle += 1
        phase = .inhale
        scheduleAdvance(after: session.inhale)
      }
    case .finished:
      break
    }
  }

  private func duration(of phase: Phase) -> TimeInterval {
    switch phase {
    case .inhale: session.inhale
    case .exhale: session.exhale
    case .finished: 0
    }
  }
}

extension BreathEngine {
  /// A stopped engine pinned to one moment of the session, for previews.
  /// `session` defaults to the balancing breath (resolved in the body — a
  /// default argument can't touch the MainActor-isolated library).
  static func still(
    session: DeepSession? = nil,
    phase: Phase,
    cycle: Int = 1,
    isPaused: Bool = false
  ) -> BreathEngine {
    let engine = BreathEngine(session: session ?? DeepSessionLibrary.balancingBreath)
    engine.phase = phase
    engine.cycle = cycle
    engine.isPaused = isPaused
    return engine
  }
}

File: android/core/model/src/main/java/io/appbeyond/freelance/deep/feature/deepsession/model/BreathEngine.kt
package io.appbeyond.freelance.deep.feature.deepsession.model

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.ZERO

/**
 * Drives one guided session: which breath phase we're in, which round, and whether
 * the practice is paused or complete.
 *
 * Pure clockwork — no UI, no audio — so the screen stays a projection of this
 * state and a preview can pin any moment of the session.
 *
 * Ported from Deep/Deep/Features/DeepSession/Models/BreathEngine.swift. The
 * structure is unchanged; two things are injected that iOS reaches for globally:
 *
 * - [scope] instead of an unstructured `Task {}`, so the engine's clock is owned
 *   by whoever created it and dies with them.
 * - [nowMillis] instead of `Date()`, which is what lets a test drive the whole
 *   session on virtual time. The iOS suite has to run against a real clock and
 *   says so in its own header: *"phases and timeouts here are kept short but
 *   generous — fast enough to run quickly, loose enough not to flake on a busy CI
 *   machine."* With a supplied clock, the Kotlin tests assert exact boundaries in
 *   microseconds and cannot flake.
 */
class BreathEngine(
  val session: DeepSession,
  private val scope: CoroutineScope,
  private val nowMillis: () -> Long = System::currentTimeMillis,
) {

  enum class Phase { Inhale, Exhale, Finished }

  data class State(
    val phase: Phase = Phase.Inhale,
    /** 1-based current round, capped at [DeepSession.cycles]. */
    val cycle: Int = 1,
    val isPaused: Boolean = false,
  )

  private val _state = MutableStateFlow(State())
  val state: StateFlow<State> = _state.asStateFlow()

  private var advanceJob: Job? = null
  private var phaseEndsAtMillis: Long? = null

  /** What was left of the current phase when it was paused. */
  private var pausedRemaining: Duration? = null

  /** Starts the clock. Safe to call once on first composition; later calls no-op. */
  fun begin() {
    val current = _state.value
    if (advanceJob != null || current.phase == Phase.Finished || current.isPaused) return
    scheduleAdvance(durationOf(current.phase))
  }

  fun togglePaused() {
    if (_state.value.phase == Phase.Finished) return
    if (_state.value.isPaused) resume() else pause()
  }

  /** Stops the clock without changing visible state; call on dismissal. */
  fun cancel() {
    advanceJob?.cancel()
    advanceJob = null
  }

  /**
   * Internal rather than only [togglePaused]'s half, so the screen can settle the
   * practice when the app leaves the foreground.
   */
  fun pause() {
    val current = _state.value
    if (current.phase == Phase.Finished || current.isPaused) return
    pausedRemaining = remainingInPhase
    cancel()
    _state.value = current.copy(isPaused = true)
  }

  private fun resume() {
    val remaining = pausedRemaining ?: durationOf(_state.value.phase)
    pausedRemaining = null
    _state.value = _state.value.copy(isPaused = false)
    scheduleAdvance(remaining)
  }

  /**
   * What is left of the current phase — live while the clock runs, frozen while
   * paused. Null before [begin] and once finished.
   */
  val remainingInPhase: Duration?
    get() = pausedRemaining
      ?: phaseEndsAtMillis?.let { end -> (end - nowMillis()).coerceAtLeast(0L).milliseconds }

  /** The full length of the phase we're currently in. */
  val currentPhaseDuration: Duration get() = durationOf(_state.value.phase)

  /**
   * Practice still ahead — what remains of this phase, the rest of this round, and
   * every round after it. Read live, like [remainingInPhase], so a caller re-reads
   * it whenever something else moves them.
   */
  val remainingInSession: Duration
    get() {
      val current = _state.value
      if (current.phase == Phase.Finished) return ZERO
      val inThisPhase = remainingInPhase ?: durationOf(current.phase)
      val restOfRound = if (current.phase == Phase.Inhale) session.exhale else ZERO
      val roundsAhead = (session.cycles - current.cycle).coerceAtLeast(0)
      return inThisPhase + restOfRound + session.cycleDuration * roundsAhead
    }

  private fun scheduleAdvance(after: Duration) {
    phaseEndsAtMillis = nowMillis() + after.inWholeMilliseconds
    advanceJob = scope.launch {
      delay(after)
      if (isActive) advance()
    }
  }

  private fun advance() {
    val current = _state.value
    when (current.phase) {
      Phase.Inhale -> {
        _state.value = current.copy(phase = Phase.Exhale)
        scheduleAdvance(session.exhale)
      }

      Phase.Exhale ->
        if (current.cycle >= session.cycles) {
          _state.value = current.copy(phase = Phase.Finished)
          phaseEndsAtMillis = null
          advanceJob = null
        } else {
          _state.value = current.copy(phase = Phase.Inhale, cycle = current.cycle + 1)
          scheduleAdvance(session.inhale)
        }

      Phase.Finished -> Unit
    }
  }

  private fun durationOf(phase: Phase): Duration = when (phase) {
    Phase.Inhale -> session.inhale
    Phase.Exhale -> session.exhale
    Phase.Finished -> ZERO
  }
}
